DESIGN_GUIDELINES.md

```diff
@@ -0,0 +1,671 @@
+- support for stdin (CONVENTION??)
+
+### Feedback on successful operations
+
+Successful operations on an object should print the object's identifier on `stdout`
+on success; doing so enables users to consume the output for scripting.
+
+For example, the following command creates two volumes, using different drivers,
+and attaches those volumes to a new container:
+
+```bash
+docker container run \
+  --volume $(docker volume create --driver=foo):/somewhere \
+  --volume $(docker volume create --driver=bar):/somewhere-else \
+  busybox
+```
+
+
+Either `ID` or `name` are acceptable, as long as the identifier can be used
+to reference the object.
+
+Some examples of commands that follow this design:
+
+
+Creating a container prints the container's `ID` on `stdout`
+
+```bash
+docker container create --name test busybox
+cc1555ede50a11f02a3ef6cc8eedd9f78f6299a2f2b7efdc8a91fbefb8fc194a
+```
+
+
+Removing a container prints the reference that was given
+
+```bash
+docker container rm test
+test
+
+docker container rm 057d35243903e522c70ae2dfab5f706e4ea3cc1ae4c38cce955a99be3d09cfd1
+057d35243903e522c70ae2dfab5f706e4ea3cc1ae4c38cce955a99be3d09cfd1
+```
+
+
+
+### Use of stdout and stderr
+
+
+As a rule of thumb;
+
+- Use `stdout` to print the _expected_ output of a command
+- Where possible make `stdout` usable for scripting
+- Use `stderr` non-standard output (errors), and for informational messages.
+
+Sometimes these differences are subtle.
+
+#### Example: usage information.
+
+The `docker` command expects a subcommand. If no subcommand is given, the Docker
+CLI prints an informational message, showing the usage information:
+
+```bash
+docker
+
+Usage:	docker [OPTIONS] COMMAND
+
+A self-sufficient runtime for containers
+...
+```
+
+Typing `docker --help` also prints the usage information:
+
+```bash
+docker --help
+
+Usage:	docker [OPTIONS] COMMAND
+
+A self-sufficient runtime for containers
+...
+```
+
+At a glance, both appear to be identical, but there is a difference:
+
+In the first example, `stdout` contains the "expected" output of the `docker`
+command; the `docker` command requires a subcommand, and by itself does not
+produce a result.
+
+The "usage" output is informational, and therefore printed on `stderr`, which
+can be seen when discarding the `stderr` output by redirecting it to `/dev/null`;
+
+```bash
+docker 2> /dev/null
+# no output (stdout contains no output)
+```
+
+When using the `--help` flag, the usage information is the _expected_ output,
+and therefore printed on `stdout`. Suppressing `stderr` output shows that all
+output is this time printed on `stdout`:
+
+```bash
+docker --help 2> /dev/null
+
+Usage:	docker [OPTIONS] COMMAND
+
+A self-sufficient runtime for containers
+```
+
+
+#### Practical example: docker run
+
+
+logging informational messages and consuming stdout output
+
+```bash
+docker container run -d nginx:alpine | xargs docker container inspect --format 'the name of the started container is: {{.Name}}'
```

Comment:
nit: Prefix the commands with a `$ ` ?

Reply:
If you did then you'd need to change the formatting from `bash` to `console` too since the section would no longer be a pure bash snippet but a console example.